File: api/app/routes/feedback.py
from fastapi import APIRouter, Depends, HTTPException, Request, status
from slowapi import Limiter
from slowapi.util import get_remote_address
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import Session

from app import metrics, models
from app.database import get_db
from app.logger import get_logger
from app.schemas import FeedbackStatusResponse, FeedbackSubmitRequest, FeedbackSubmitResponse
from app.security import require_session
from app.services.credits import add_credits, get_credit_balance
from app.services.feedback_settings import (
    FEEDBACK_CAMPAIGN,
    FEEDBACK_REWARD_TOKENS,
    feedback_is_eligible,
)

log = get_logger("carver.feedback")
router = APIRouter(prefix="/feedback", tags=["feedback"])
_limiter = Limiter(key_func=get_remote_address)


@router.get("/status", response_model=FeedbackStatusResponse)
def feedback_status(
    source: str = "website_popup",
    session: dict = Depends(require_session),
    db: Session = Depends(get_db),
):
    user_key = session.get("sub", "")
    safe_source = "whatsapp_message" if source == "whatsapp_message" else "website_popup"
    eligible, setting = feedback_is_eligible(db, user_key=user_key, source=safe_source)
    submitted = (
        db.query(models.FeedbackSubmission.id)
        .filter(
            models.FeedbackSubmission.user_key == user_key,
            models.FeedbackSubmission.campaign == FEEDBACK_CAMPAIGN,
        )
        .first()
    ) is not None
    return {
        "ok": True,
        "submitted": submitted,
        "eligible": eligible,
        "force_prompt": eligible and setting.target_mode == "specific",
        "target_mode": setting.target_mode,
        "reward_amount": FEEDBACK_REWARD_TOKENS,
    }


@router.post("/submit", response_model=FeedbackSubmitResponse, status_code=status.HTTP_201_CREATED)
@_limiter.limit("5/minute")
def submit_feedback(
    request: Request,
    payload: FeedbackSubmitRequest,
    session: dict = Depends(require_session),
    db: Session = Depends(get_db),
):
    user_key = session.get("sub", "")
    if not user_key:
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Authentication required.")

    eligible, _setting = feedback_is_eligible(db, user_key=user_key, source=payload.source)
    if not eligible:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Feedback reward is not available for this account right now.",
        )

    existing = (
        db.query(models.FeedbackSubmission.id)
        .filter(
            models.FeedbackSubmission.user_key == user_key,
            models.FeedbackSubmission.campaign == FEEDBACK_CAMPAIGN,
        )
        .first()
    )
    if existing:
        return {
            "ok": True,
            "submitted": True,
            "reward_granted": False,
            "reward_amount": FEEDBACK_REWARD_TOKENS,
            "credits_balance": get_credit_balance(db, user_key),
        }

    submission = models.FeedbackSubmission(
        user_key=user_key,
        campaign=FEEDBACK_CAMPAIGN,
        source=payload.source,
        rating=payload.rating,
        liked=payload.liked,
        improved=payload.improved,
        confusing=payload.confusing,
        recommend=payload.recommend,
        anything_else=payload.anything_else,
        reward_amount=FEEDBACK_REWARD_TOKENS,
        rewarded=FEEDBACK_REWARD_TOKENS > 0,
    )
    db.add(submission)
    try:
        db.flush()
    except IntegrityError:
        db.rollback()
        return {
            "ok": True,
            "submitted": True,
            "reward_granted": False,
            "reward_amount": FEEDBACK_REWARD_TOKENS,
            "credits_balance": get_credit_balance(db, user_key),
        }

    if FEEDBACK_REWARD_TOKENS > 0:
        credits_balance = add_credits(db, user_key, amount=FEEDBACK_REWARD_TOKENS)
    else:
        db.commit()
        credits_balance = get_credit_balance(db, user_key)
    metrics.increment("feedback_submissions")
    log.info(
        "Feedback submitted | user=%s | source=%s | reward=%d",
        user_key,
        payload.source,
        FEEDBACK_REWARD_TOKENS,
    )
    return {
        "ok": True,
        "submitted": True,
        "reward_granted": FEEDBACK_REWARD_TOKENS > 0,
        "reward_amount": FEEDBACK_REWARD_TOKENS,
        "credits_balance": credits_balance,
    }

File: jobcarver-go/internal/handlers/feedback/feedback.go
// Package feedback ports api/app/routes/feedback.py plus the relevant pieces of
// app/services/feedback_settings.py (campaign settings + eligibility) so the
// handler stays self-contained.
package feedback

import (
	"encoding/json"
	"net/http"
	"strings"
	"time"

	"github.com/go-chi/chi/v5"

	"jobcarver/internal/config"
	"jobcarver/internal/credits"
	"jobcarver/internal/db"
	"jobcarver/internal/httpx"
	"jobcarver/internal/metrics"
	"jobcarver/internal/models"
	"jobcarver/internal/security"
)

const (
	feedbackCampaign     = "feedback_2_tokens"
	feedbackRewardTokens = 0 // reward currently disabled; submissions still recorded
)

var validTargetModes = map[string]bool{"all": true, "website": true, "whatsapp": true, "specific": true, "off": true}

// Register mounts /feedback (all routes require a session).
func Register(r chi.Router) {
	r.Route("/feedback", func(r chi.Router) {
		r.Use(security.RequireSession)
		r.Get("/status", status)
		r.Post("/submit", submit)
	})
}

func status(w http.ResponseWriter, r *http.Request) {
	sess := security.SessionFromContext(r.Context())
	userKey := sess.Sub

	source := r.URL.Query().Get("source")
	safeSource := "website_popup"
	if source == "whatsapp_message" {
		safeSource = "whatsapp_message"
	}

	setting := getFeedbackSetting()
	eligible := feedbackIsEligible(setting, userKey, safeSource)

	var cnt int64
	db.DB.Model(&models.FeedbackSubmission{}).
		Where("user_key = ? AND campaign = ?", userKey, feedbackCampaign).Count(&cnt)

	httpx.JSON(w, http.StatusOK, map[string]any{
		"ok":            true,
		"submitted":     cnt > 0,
		"eligible":      eligible,
		"force_prompt":  eligible && setting.TargetMode == "specific",
		"target_mode":   setting.TargetMode,
		"reward_amount": feedbackRewardTokens,
	})
}

type submitRequest struct {
	Source       string  `json:"source"`
	Rating       int     `json:"rating"`
	Liked        *string `json:"liked"`
	Improved     *string `json:"improved"`
	Confusing    *string `json:"confusing"`
	Recommend    *string `json:"recommend"`
	AnythingElse *string `json:"anything_else"`
}

func submit(w http.ResponseWriter, r *http.Request) {
	sess := security.SessionFromContext(r.Context())
	userKey := sess.Sub
	if userKey == "" {
		httpx.WriteError(w, http.StatusUnauthorized, "Authentication required.", httpx.CRV2004)
		return
	}

	var body submitRequest
	if err := httpx.Decode(r, &body); err != nil {
		httpx.WriteError(w, http.StatusUnprocessableEntity, "Invalid request body.", httpx.CRV1003)
		return
	}
	if body.Source == "" {
		body.Source = "website_popup"
	}
	if body.Source != "website_popup" && body.Source != "whatsapp_message" {
		httpx.WriteError(w, http.StatusUnprocessableEntity, "Invalid source.", httpx.CRV1003)
		return
	}
	if body.Rating < 1 || body.Rating > 5 {
		httpx.WriteError(w, http.StatusUnprocessableEntity, "rating must be between 1 and 5.", httpx.CRV1003)
		return
	}
	if body.Recommend != nil {
		switch *body.Recommend {
		case "yes", "no", "maybe":
		default:
			httpx.WriteError(w, http.StatusUnprocessableEntity, "Invalid recommend value.", httpx.CRV1003)
			return
		}
	}
	stripOptional(&body.Liked)
	stripOptional(&body.Improved)
	stripOptional(&body.Confusing)
	stripOptional(&body.AnythingElse)

	setting := getFeedbackSetting()
	if !feedbackIsEligible(setting, userKey, body.Source) {
		httpx.WriteError(w, http.StatusForbidden,
			"Feedback reward is not available for this account right now.", httpx.CRV2005)
		return
	}

	var existing int64
	db.DB.Model(&models.FeedbackSubmission{}).
		Where("user_key = ? AND campaign = ?", userKey, feedbackCampaign).Count(&existing)
	if existing > 0 {
		httpx.JSON(w, http.StatusCreated, map[string]any{
			"ok":              true,
			"submitted":       true,
			"reward_granted":  false,
			"reward_amount":   feedbackRewardTokens,
			"credits_balance": credits.GetBalance(db.DB, userKey),
		})
		return
	}

	submission := models.FeedbackSubmission{
		UserKey:      userKey,
		Campaign:     feedbackCampaign,
		Source:       body.Source,
		Rating:       body.Rating,
		Liked:        body.Liked,
		Improved:     body.Improved,
		Confusing:    body.Confusing,
		Recommend:    body.Recommend,
		AnythingElse: body.AnythingElse,
		RewardAmount: feedbackRewardTokens,
		Rewarded:     feedbackRewardTokens > 0,
	}
	if err := db.DB.Create(&submission).Error; err != nil {
		// Likely the (user_key, campaign) unique constraint — treat as already submitted.
		httpx.JSON(w, http.StatusCreated, map[string]any{
			"ok":              true,
			"submitted":       true,
			"reward_granted":  false,
			"reward_amount":   feedbackRewardTokens,
			"credits_balance": credits.GetBalance(db.DB, userKey),
		})
		return
	}

	if feedbackRewardTokens > 0 {
		_ = credits.Grant(db.DB, userKey, feedbackRewardTokens)
	}
	balance := credits.GetBalance(db.DB, userKey)
	metrics.Increment("feedback_submissions")
	httpx.JSON(w, http.StatusCreated, map[string]any{
		"ok":              true,
		"submitted":       true,
		"reward_granted":  feedbackRewardTokens > 0,
		"reward_amount":   feedbackRewardTokens,
		"credits_balance": balance,
	})
}

func stripOptional(p **string) {
	if *p == nil {
		return
	}
	s := strings.TrimSpace(**p)
	if s == "" {
		*p = nil
		return
	}
	*p = &s
}

// getFeedbackSetting mirrors feedback_settings.get_feedback_setting: load (or
// lazily create) the single campaign row, normalising an invalid target_mode.
func getFeedbackSetting() models.FeedbackCampaignSetting {
	var setting models.FeedbackCampaignSetting
	err := db.DB.Where("campaign = ?", feedbackCampaign).First(&setting).Error
	if err == nil {
		if !validTargetModes[setting.TargetMode] {
			setting.TargetMode = "all"
			db.DB.Save(&setting)
		}
		return setting
	}
	setting = models.FeedbackCampaignSetting{
		Campaign:           feedbackCampaign,
		Enabled:            true,
		TargetMode:         "all",
		TargetUserKeysJSON: "[]",
	}
	db.DB.Create(&setting)
	return setting
}

func feedbackIsEligible(setting models.FeedbackCampaignSetting, userKey, source string) bool {
	mode := setting.TargetMode
	if !validTargetModes[mode] {
		mode = "all"
	}
	if !setting.Enabled || mode == "off" {
		return false
	}

	var modeEligible bool
	switch mode {
	case "all":
		modeEligible = true
	case "website":
		modeEligible = source == "website_popup"
	case "whatsapp":
		modeEligible = source == "whatsapp_message"
	case "specific":
		targets := parseTargets(setting.TargetUserKeysJSON)
		modeEligible = targets[normaliseUserKey(userKey)]
	}
	if !modeEligible {
		return false
	}
	if source == "website_popup" && !userOldEnoughForWebsiteFeedback(userKey) {
		return false
	}
	return true
}

func userOldEnoughForWebsiteFeedback(userKey string) bool {
	minAgeHours := config.S.FeedbackMinAccountAgeHours
	if minAgeHours <= 0 {
		return true
	}
	key := normaliseUserKey(userKey)
	if key == "" {
		return false
	}
	var user models.User
	if err := db.DB.Select("created_at").Where("email = ?", key).First(&user).Error; err != nil {
		// No website user row (admin/phone identities) — keep compatible.
		return true
	}
	age := time.Since(user.CreatedAt)
	return age >= time.Duration(minAgeHours)*time.Hour
}

func parseTargets(raw string) map[string]bool {
	out := map[string]bool{}
	var items []string
	if json.Unmarshal([]byte(raw), &items) != nil {
		return out
	}
	for _, it := range items {
		if k := normaliseUserKey(it); k != "" {
			out[k] = true
		}
	}
	return out
}

func normaliseUserKey(k string) string {
	return strings.ToLower(strings.TrimSpace(k))
}
